test.sh

```diff
@@ -0,0 +1,32 @@
+#!/bin/bash
+set -o errexit -o nounset -o pipefail
+cd "$(dirname "$0")"
+
+msg() {
+  echo "$@" >&2
+}
+
+cmd() {
+  echo "Running $*" >&2
+  "$@"
+}
+
+msg 'TEST: no features'
+cmd cargo test --no-default-features
+cmd cargo clippy --no-default-features
+cmd env RUSTDOCFLAGS='-D warnings' cargo doc --no-default-features
+
+msg 'TEST: std'
+cmd cargo test --no-default-features --features=std
+cmd cargo clippy --no-default-features --features=std
+cmd env RUSTDOCFLAGS='-D warnings' cargo doc --no-default-features --features=std
+
+msg 'TEST: serde'
+cmd cargo test --no-default-features --features=serde
+cmd cargo clippy --no-default-features --features=serde
+cmd env RUSTDOCFLAGS='-D warnings' cargo doc --no-default-features --features=serde
```

Comment:
Is there a more elegant way of testing these?

Reply:
> Is there a more elegant way of testing these?

I suppose you can write a script that reads `Cargo.toml` and create every feature combination and loop over them? But that would be overly complicated. Our time is better spent on the code of the library itself.

We can also reduce the number of combination by making `serde` requires `std`, or merging `serde` into `std`.

(BTW, `--all-features` does not replace this script, `--all-features` is but one of the possible combinations).

But if what you meant is you don't like `.sh` file then we can switch to `justfile` or `Makefile` (preferably on another PR).

Reply:
Do you want to use this tool? https://github.com/frewsxcv/cargo-all-features

Reply:
Can you move this file to `.github` folder and rename it to something like `test-all-features.sh`?

Reply:
Also `cargo hack` supports this https://github.com/taiki-e/cargo-hack#--feature-powerset

Reply:
> Can you move this file to `.github` folder and rename it to something like `test-all-features.sh`?

This script is not actually used by the CI. It is used by me in the terminal.

> Also `cargo hack` supports this https://github.com/taiki-e/cargo-hack#--feature-powerset

Does it support doc checking and clippy linting?